docs/csharp/tutorials/upgrade-to-nullable-references.md

```diff
@@ -0,0 +1,153 @@
+
+The sample application that you'll migrate is an RSS feed reader web app. It reads from a single RSS feed and displays summaries for the most recent articles. You can click on any of the articles to visit the site. The application is relatively new, but was written before nullable reference types were available. The design decisions for the application represented sound principles, but don't take advantage of this important language feature.
+
+The sample application includes a unit test library that validates the major functionality of the app. That project will make it easier to upgrade safely, if you change any of the implementation based on the warnings generated. You can download the starter code from our [samples GitHub repo](https://github.com/dotnet/samples/tree/master/csharp/tutorials/nullable-reference-migration/start).
+
+Your goal migrating a project should be to leverage the new language features so that you clearly express your intent on the nullability of variables, and do so in such a way that the  compiler does not generate warnings when you have the nullable annotation context and nullable warning context set to `enabled`.
+
+## Upgrade the projects to C# 8
+
+A good first step is to determine the scope of the migration task. Start by upgrading the project to C# 8.0 (or newer). Add the `LangVersion` element to both csproj files for the web project and the unit test project:
+
+```xml
+<LangVersion>8.0</LangVersion>
+```
+
+Upgrading the language version selects C# 8.0, but does not enable the nullable annotation context or the nullable warning context. Rebuild the project to ensure that it builds without warnings.
+
+A good next step is to turn on the nullable annotation context and see how many warnings are generated.  Add the following element to both csproj files in the solution, directly under the `LangVersion` element:
+
+```xml
+<NullableContextOptions>enable</NullableContextOptions>
+```
+
+Do a test build, and notice the warning list. In this small application, the compiler generates five warnings, so it's likely you'd leave the nullable annotation context enabled and start fixing warnings for the entire project.
+
+That strategy works only for the smallest projects. For any larger projects, the number of warnings generated by enabling the nullable annotation context for the entire codebase makes it harder to fix the warnings systematically.
+
+## Warnings help discover original design intent
+
+There are two classes that generate multiple warnings. Start with the `NewsStoryViewModel` class. Remove the `NullableContextOptions` element from both csproj files so that you can limit the scope of warnings to the sections of code you're working with. Open the `NewsStoryViewModel.cs` and add the following directives to enable the nullable annotation context for the `NewsStoryViewModel` and restore it following that class definition:
```

Comment:
```suggestion
There are two classes that generate multiple warnings. Start with the `NewsStoryViewModel` class. Remove the `NullableContextOptions` element from both csproj files so that you can limit the scope of warnings to the sections of code you're working with. Open the *NewsStoryViewModel.cs* file and add the following directives to enable the nullable annotation context for the `NewsStoryViewModel` and restore it following that class definition:
```